### PR TITLE
Extract `HistoricalView` trait from the `AtomicView`

### PR DESCRIPTION
**Pull request by xgreenx**
It is preparation for https://github.com/FuelLabs/fuel-core/issues/451.

Most of the services don't need historical functionality; only the executor uses the view of the storage at a specific height.

The change extracts this functionality into its own `HistoricalView` trait.

During the implementation of the state rewind, I realized that the historical view has some limitations(like we are not able to iterate). So because of that, the historical view uses another type that only implements the `KeyValueInspect` trait.

## Checklist
- [x] Breaking changes are clearly marked as such in the PR description and changelog

### Before requesting review
- [x] I have reviewed the code myself

### CODE REVIEW

File: bin/fuel-core/chainspec/local-testnet/state_transition_bytecode.wasm
Comment (Dentosal) [inline bin/fuel-core/chainspec/local-testnet/state_transition_bytecode.wasm:1]: Why is this removed?

Comment (xgreenx) [reply bin/fuel-core/chainspec/local-testnet/state_transition_bytecode.wasm:1]: It is not removed=) Because of simplification in the trait, the compiler removed 191 bytes from the bytecode. Bu the file still exists

Comment (Dentosal) [reply bin/fuel-core/chainspec/local-testnet/state_transition_bytecode.wasm:1]: image

GitHub says -100% which makes me thing the whole file is removed, but apparently this is not the case.

File: crates/fuel-core/src/graphql_api/view_extension.rs [hunk 1]
```diff
@@ -37,7 +39,15 @@ impl Extension for ViewExtension {
...
     ) -> ServerResult<Request> {
         let database: &ReadDatabase = ctx.data_unchecked();
-        let view = database.view();
+        let view = database.view().map_err(|e| {
+            ServerError::new(
+                e.to_string(),
+                Some(Pos {
+                    line: line!() as usize,
+                    column: column!() as usize,
+                }),
```
Comment (Dentosal) [inline crates/fuel-core/src/graphql_api/view_extension.rs:48]: This feels weird. Maybe it's actually supposed to work like this, but the mismatch between `line!` and `column!` evaluated separately seems wrong.

File: bin/fuel-core/chainspec/local-testnet/state_transition_bytecode.wasm
Comment (xgreenx) [reply bin/fuel-core/chainspec/local-testnet/state_transition_bytecode.wasm:1]: Yeah, it is the UI bug of the girhub=(

File: crates/storage/src/transactional.rs [hunk 2]
```diff
@@ -33,23 +33,29 @@ use crate::{
...
-    /// Returns the view of the storage at the given `height`.
-    fn view_at(&self, height: &Self::Height) -> StorageResult<Self::View>;
-
     /// Returns the view of the storage for the latest block height.
-    fn latest_view(&self) -> Self::View;
+    fn latest_view(&self) -> StorageResult<Self::LatestView>;
+}
+
+/// Provides a view of the storage at the given height.
+pub trait HistoricalView: AtomicView {
```
Comment (Dentosal) [inline crates/storage/src/transactional.rs:53]: Since `HistoricalView` is a subtrait of `AtomicView`, it'd like that the comment here (and possibly on `AtomicView` as well) would mention that `AtomicView` doesn't mean the latest version here.

File: crates/storage/src/transactional.rs [hunk 3]
```diff
@@ -33,23 +33,29 @@ use crate::{
     kv_store::KVItem,
 };
 
-/// Provides a view of the storage at the given height.
+/// Provides a view of the storage at the latest height.
 /// It guarantees to be atomic, meaning the view is immutable to outside modifications.
 pub trait AtomicView: Send + Sync {
```
Comment (Dentosal) [inline crates/storage/src/transactional.rs:38]: I find the wording/naming to be unclear. How can it provide a view of the latest height, but then not change when the latest height changes?

Comment (xgreenx) [reply crates/storage/src/transactional.rs]: I also decided to move the `latest_height` to `HistoricalView` since it looks like a better place for this logic.

**Review by Dentosal — APPROVED** [review]
